components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttAsyncCallback.java

```diff
@@ -0,0 +1,285 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.inbound.endpoint.protocol.mqtt;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.eclipse.paho.client.mqttv3.*;
+
+/**
+ * MQTT Asynchronous call back handler
+ */
+public class MqttAsyncCallback implements MqttCallback {
+
+    //state variables
+    static final int BEGIN = 0;
+    static final int CONNECTED = 1;
+    static final int PUBLISHED = 2;
+    static final int SUBSCRIBED = 3;
+    static final int DISCONNECTED = 4;
+    static final int FINISH = 5;
+    static final int ERROR = 6;
+    static final int DISCONNECT = 7;
+    private final Object waiter = new Object();
+    int state = BEGIN;
+    private MqttConnectOptions conOpt;
+    private Log log = LogFactory.getLog(MqttAsyncCallback.class);
+    // Private instance variables
+    private MqttAsyncClient client;
+    private Throwable ex = null;
+    private boolean donext = false;
+
+    private MqttListener asycClient;
+
+    private MqttInjectHandler injectHandler;
+
+    public MqttAsyncCallback(MqttAsyncClient clientAsync, MqttInjectHandler injectHandler) throws MqttException {
+
+        client = clientAsync;
+        // Set this wrapper as the callback handler
+        client.setCallback(this);
+
+        this.injectHandler = injectHandler;
+
+    }
+
+    public void setConOpt(MqttConnectOptions conOpt) {
+        this.conOpt = conOpt;
+    }
+
+    /**
+     * Wait for a maximum amount of time for a state change event to occur
+     *
+     * @param maxTTW maximum time to wait in milliseconds
+     * @throws MqttException
+     */
+    private void waitForStateChange(int maxTTW) throws MqttException {
+        synchronized (waiter) {
+            if (!donext) {
+                try {
+                    waiter.wait(maxTTW);
+                } catch (InterruptedException e) {
+                    log.error("timed out");//TODO LOG.ERROR
+                }
+
+            }
+            donext = false;
+        }
+    }
+
+    /**
+     * Subscribe to a topic on an MQTT server Once subscribed this method waits
+     * for the messages to arrive from the server that match the subscription.
+     *
+     * @param topicName to subscribe to (can be wild carded)
+     * @param qos       the maximum quality of service to receive messages at for this
+     *                  subscription
+     * @throws MqttException
+     */
+    public void subscribe(String topicName, int qos) throws Throwable {
+        // Use a state machine to decide which step to do next. State change
+        // occurs
+        // when a notification is received that an MQTT action has completed
+        while (state != FINISH) {
+            switch (state) {
+                case BEGIN:
```

Comment:
We discussed that we don't need to wait for connect method. What are we still waiting for that?